e2e/playwright/editorFixture.ts

```diff
@@ -0,0 +1,109 @@
+import type { Page, Locator } from '@playwright/test'
+import { expect } from '@playwright/test'
+
+interface EditorState {
+  activeLines: Array<string>
+  highlightedCode: string
+  diagnostics: Array<string>
+}
+
+function removeWhitespace(str: string) {
+  return str.replace(/\s+/g, '').trim()
+}
+export class EditorFixture {
+  public readonly page: Page
+
+  private readonly diagnosticsTooltip: Locator
+  private readonly diagnosticsGutterIcon: Locator
+  private readonly codeContent: Locator
+  private readonly activeLine: Locator
+
+  constructor(page: Page) {
+    this.page = page
+
+    this.codeContent = page.locator('.cm-content')
+    this.diagnosticsTooltip = page.locator('.cm-tooltip-lint')
+    this.diagnosticsGutterIcon = page.locator('.cm-lint-marker-error')
+    this.activeLine = this.page.locator('.cm-activeLine')
+  }
+
+  private _expectEditorToContain =
+    (not = false) =>
+    (
+      code: string,
+      {
+        shouldNormalise = false,
+        timeout = 5_000,
+      }: { shouldNormalise?: boolean; timeout?: number } = {}
+    ) => {
+      if (!shouldNormalise) {
+        const expectStart = expect(this.codeContent)
+        if (not) {
+          return expectStart.not.toContainText(code, { timeout })
+        }
+        return expectStart.toContainText(code, { timeout })
+      }
+      const normalisedCode = code.replaceAll(/\s+/g, '').trim()
+      const expectStart = expect.poll(
+        async () => {
+          const editorText = await this.codeContent.textContent()
+          return editorText?.replaceAll(/\s+/g, '').trim()
+        },
+        {
+          timeout,
+        }
+      )
+      if (not) {
+        return expectStart.not.toContain(normalisedCode)
+      }
+      return expectStart.toContain(normalisedCode)
+    }
+  expectEditor = {
+    toContain: this._expectEditorToContain(),
+    not: { toContain: this._expectEditorToContain(true) },
+  }
+  private _serialiseDiagnostics = async (): Promise<Array<string>> => {
+    const diagnostics = await this.diagnosticsGutterIcon.all()
+    const diagnosticsContent: string[] = []
+    for (const diag of diagnostics) {
+      await diag.hover()
+      const content = await this.diagnosticsTooltip.allTextContents()
+      diagnosticsContent.push(content.join(''))
+    }
+    return [...new Set(diagnosticsContent)].map((d) => d.trim())
+  }
```

Comment:
❤️ 

Will probably need tweaking in future, I've seen errors that have uuids in them, we probably want to avoid showing those to users in the first place, but in the short term I can also seeing us auto detecting uuids in this code in text and changing them to `0000-0000-0000` so we have stable assertions.